src/tensorflow.cc

```diff
@@ -1332,62 +1340,264 @@ AutoCompleteHelper::FixBatchingSupport()
+      RETURN_IF_ERROR(config.Add("reshape", std::move(reshape)));
+    }
+
+    if (found_dims) {
+      RETURN_IF_ERROR(config.Remove("dims"));
+    }
+    RETURN_IF_ERROR(config.Add("dims", std::move(dims)));
+  }
+
+  // Elements in dims should match 'rank'. However, ragged batching is an
+  // exception to this rule. A tensor allowing ragged batch should not match
+  // with 'rank - 1'.
+  if (!using_ragged_batching_) {
+    triton::common::TritonJson::Value current_dims(
+        model_state_->ModelConfig(),
+        triton::common::TritonJson::ValueType::ARRAY);
+    config.Find("dims", &current_dims);
+
+    if (model_support_batching_) {
+      RETURN_ERROR_IF_TRUE(
+          current_dims.ArraySize() != (io->shape_->rank_ - 1),
+          TRITONSERVER_ERROR_INVALID_ARG,
+          std::string(
+              "Number of dimensions (" +
+              std::to_string(current_dims.ArraySize()) + ") given for tensor " +
+              io->name_ + " for model '" + model_state_->Name() +
+              "' in configuration does not match the rank (" +
+              std::to_string(io->shape_->rank_ - 1) +
+              ") of the loaded model."));
+    } else {
+      RETURN_ERROR_IF_TRUE(
+          current_dims.ArraySize() != io->shape_->rank_,
+          TRITONSERVER_ERROR_INVALID_ARG,
+          std::string(
+              "Number of dimensions (" +
+              std::to_string(current_dims.ArraySize()) + ") given for tensor " +
+              io->name_ + " for model '" + model_state_->Name() +
+              "' in configuration does not match the rank (" +
+              std::to_string(io->shape_->rank_) + ") of the loaded model."));
+    }
+  }
+
+  return nullptr;  // Success
+}
+
+void
+AutoCompleteHelper::RemoveFromListByName(
+    const char* name, std::vector<const TRITONTF_IOList*>& list)
+{
+  for (size_t i = 0; i < list.size(); ++i) {
+    const TRITONTF_IOList* current = list.at(i);
+    if (!strcmp(name, current->io_->name_)) {
+      list.erase(list.begin() + i);
+      return;
+    }
+  }
+}
+
+std::vector<const TRITONTF_IOList*>
+AutoCompleteHelper::CopyList(const TRITONTF_IOList* list)
+{
+  std::vector<const TRITONTF_IOList*> copy;
+  if (list == nullptr) {
+    return copy;
+  }
 
-  triton::common::TritonJson::Value auto_complete_ios(
+  while (list != nullptr) {
+    copy.push_back(list);
+    list = list->next_;
+  }
+
+  return copy;
+}
+
+TRITONSERVER_Error*
+AutoCompleteHelper::FixIOConfigInputs(const TRITONTF_IOList* reference_list)
+{
+  triton::common::TritonJson::Value ios;
+  model_state_->ModelConfig().Find("input", &ios);
+
+  // Input names must all be defined and match the user provided config.
+  // - If there is an input from the model which is not in config.pbtxt
+  //   then we try to autocomplete it.
+  // - If there is an input in the config.pbtxt
+  //   which is not in the model then we throw an error.
+  std::vector<const TRITONTF_IOList*> reference_list_copy =
+      CopyList(reference_list);
```

Comment:
This is more of a style thing. Let me know if I should create a copy of the linked list with pointers instead of this vector. Benefits would include deleting a variation of the FindModelIOByName() function. Disadvantage might be readability.

Reply:
Fixed